File: 2023/day_02/1.py
#!/usr/bin/env python3

import re

expression = "^Game ([0-9]+):(.*)$"
parser = re.compile(expression)

games = { }

# games structure :
# {
#   1: [ { red: x, green: y,blue: z }, ...],
#   ...
# }

maxcolors = { 'red': 12, 'green': 13, 'blue':14 }

with open('./input.txt') as file:
    lines = file.readlines()
    for line in lines:
        game, results = parser.match(line).groups()
        sequences = []
        for seq in results.split(";"): # sequence
            sequence = { "red": 0, "green": 0, "blue": 0}
            for color in seq.split(","): # ball ammount
                scorecolor = re.match(" ([0-9]+) (red|green|blue)", color).groups()
                sequence[scorecolor[1]] = int(scorecolor[0])
            sequences += [sequence]
        games[int(game)] = sequences

sum = 0
for game, plays in games.items():
    valid = True
    for play in plays:
        for color in maxcolors:
            if play[color] > maxcolors[color] : valid = False
    if valid : sum += int(game)

print(sum)

File: 2024/day_02/1.jl
#!/usr/bin/env julia 

lines = []

open("./input.txt") do file
  for line in eachline(file)
    push!(lines,line)
  end
end

lines = map(line -> map(n -> parse(Int,n), split(line)), lines)

check_safe = map(line -> reduce(function((x, sign, check),y) 
    if x == -1 return (y, 0, true) end
    if check == false return (x, sign, check) end

    diff = y-x
    if abs(diff) < 1 || abs(diff) > 3 
      return (x, sign, false)
    end
   
    if sign == 0
      return (y, diff/abs(diff), true)
    end

    if sign != diff/abs(diff)
      return (x, sign, false)
    end

    return (y, diff/abs(diff), true)
  end, line, init=(-1, 0, true)), lines)

check_safe = map(check -> check[3], check_safe)

println(sum(check_safe))
